include/zephyr/kernel.h

```diff
@@ -6153,6 +6153,186 @@ void k_sys_runtime_stats_enable(void);
  */
 void k_sys_runtime_stats_disable(void);
 
+/**
+ * @defgroup k_sig Kernel Signal APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief Process ID
+ *
+ * @note Process IDs are notional at this time, as processes are unsupported in Zephyr.
+ */
+typedef k_tid_t k_pid_t;
+
+/** @brief Signal Values */
+union k_sig_val {
+	int sival_int;
+	void *sival_ptr;
+};
+
+/** @brief Signal info */
+struct k_sig_info {
+	int si_signo;
+	int si_code;
+	union k_sig_val si_value;
+};
+
+/** @brief A bitset large enough to contain all signal bits. */
+struct k_sig_set;
+
+#define K_SIG_BLOCK   0
+#define K_SIG_SETMASK 1
+#define K_SIG_UNBLOCK 2
+
+/**
+ * @brief Add a signal @a sig to signal set @a set.
+ *
+ * @param set the set to which @a sig should be added.
+ * @param sig the signal which should be added to @a set.
+ *
+ * @retval 0 on success.
+ * @return -EINVAL if @a sig is invalid.
+ */
+int k_sig_addset(struct k_sig_set *set, int sig);
+
+/**
+ * @brief Remove a signal @a sig from signal set @a set.
+ *
+ * @param set the set from which @a sig should be removed.
+ * @param sig the signal which should be removed from @a set.
+ *
+ * @retval 0 on success.
+ * @return -EINVAL if @a sig is invalid.
+ */
+int k_sig_delset(struct k_sig_set *set, int sig);
+
+/**
+ * @brief Check if signal @a sig is part of signal set @a set.
+ *
+ * @param set the set to check.
+ * @param sig the signal which should be checked.
+ *
+ * @retval 0 if @a sig is not part of @a set.
+ * @retval 1 if @a sig is part of @a set.
+ * @return -EINVAL if @a sig is invalid.
+ */
+int k_sig_ismember(const struct k_sig_set *set, int sig);
+
+/**
+ * @brief Initialize a signal set with the set of all signals.
+ *
+ * @param set signal set to initialize.
+ *
+ * @retval 0 on success.
+ * @return -EINVAL if @a sig is invalid.
+ */
+int k_sig_fillset(struct k_sig_set *set);
+
+/**
+ * @brief Initialize a signal set with the set of all signals.
+ *
+ * @param set signal set to initialize.
+ *
+ * @retval 0 on success.
+ * @return -EINVAL if @a sig is invalid.
+ */
+int k_sig_emptyset(struct k_sig_set *set);
+
+/**
+ * @brief Block signals in the accompanying @ref k_sig_set
+ *
+ * @see @ref k_sig_mask
+ */
+#define K_SIG_BLOCK 0
+
+/**
+ * @brief Set signals in the accompanying @ref k_sig_set
+ *
+ * @see @ref k_sig_mask
+ */
+#define K_SIG_SETMASK 1
+
+/**
+ * @brief Unblock signals in the accompanying @ref k_sig_set
+ *
+ * @see @ref k_sig_mask
+ */
+#define K_SIG_UNBLOCK 2
+
+/** @brief Minimum realtime signal number */
+#define K_SIG_RTMIN 32
```

Comment:
There is actually a really amazing reason *not* to define `K_SIG_RTMAX` (or `K_SIRGTMAX`).

As mentioned in the POSIX spec, `[min, max]` is inclusive. Also mentioned elsewhere, is that developers are strongly encouraged to only ever use `min + i` when looping over RT sets.

That is because, max is defined in terms of min and the number of signals (`max = min + N`).

Intuitively, the way that one would normally go about writing an inclusive for loop is with `for (i= min; i <= max; i++)`. However, if there are zero RT signals, max is still defined as `min + N = min`. The loop is evaluated (a bug).

Now, if max is not defined, then the usual approach for writing a for loop would be `for (i = 0; i < N; i++) .. x = (min + i)..` or `for (i = min; i < min + N; i++)` .  if `N = 0`, the loop is not evaluated (no bug).

I actually just debugged all of this so that others wouldn't have to later 😅

Reply:
Makes sense now! Then maybe add that note in the code/comments so that some smart developer in the future won't even attempt "improving" that implementation?

Reply:
Fixed!